File: ca341/assignment2/compare.hs
-- type definition for a binary tree that will be used throughout the program
data BinTree t = Empty | Root t (BinTree t) (BinTree t)
                deriving(Ord, Eq, Show)

insert :: (Ord a) => a -> BinTree a -> BinTree a
-- insert takes an element and a tree and returns a tree with the element inserted
insert x Empty = Root x Empty Empty -- if the tree is empty, insert the element at the root
insert x (Root y left right) -- if the tree is not empty, compare the element to the root
    | x < y = Root y (insert x left) right
    | otherwise = Root y left (insert x right)

search :: (Eq a, Ord a) => a -> BinTree a -> Bool
-- search takes an element and a tree and returns True if the element is in the tree
search x Empty = False -- if the tree is empty, the element is not in the tree
search x (Root y left right) -- if the tree is not empty, compare the element to the root
    | x == y = True -- if the element is equal to the root, the element is in the tree
    -- if the element is less than the root, search the left subtree, otherwise search the right subtree
    | x < y = search x left
    | x >= y = search x right
    | otherwise = False -- if the element is not found in the left or right subtree, the element is not in the tree

preorder :: BinTree t -> [t]
-- preorder takes a tree and returns a list of the elements in the tree in preorder traversal
preorder Empty = [] -- if the tree is empty, return an empty list
preorder (Root x left right) = [x] ++ preorder left ++ preorder right -- if the tree is not empty, return the root then recursively call preorder on the left and right subtrees

postorder :: BinTree t -> [t]
-- postorder takes a tree and returns a list of the elements in the tree in postorder traversal
postorder Empty = [] -- if the tree is empty, return an empty list
postorder (Root x left right) = postorder left ++ postorder right ++ [x] -- if the tree is not empty, recursively call postorder on the left and right subtrees then return the root

inorder :: BinTree t -> [t]
-- inorder takes a tree and returns a list of the elements in the tree in inorder traversal
inorder Empty = [] -- if the tree is empty, return an empty list
inorder (Root x left right) = inorder left ++ [x] ++ inorder right -- if the tree is not empty, recursively call inorder on the left subtree then return the root then recursively call inorder on the right subtree

File: ca341/assignment2/compare.pl
% Format for the binary tree used throughout the program
% t(Root, Left, Right)

% insert a new node into a binary tree and return the new tree
insert(X, nil, t(X, nil, nil)) :- !. % if the tree is empty, create a new tree with the new node as the root
insert(X, t(Y, L, R), t(Y, L1, R)) :- X < Y, insert(X, L, L1). % insert into left subtree if X is less than Y
insert(X, t(Y, L, R), t(Y, L, R1)) :- X >= Y, insert(X, R, R1). % insert into right subtree if X is greater than or equal to Y

% search for a node in a binary tree
search(X, t(X, _, _)) :- !. % if X is the root, return true
search(X, t(Y, L, _)) :- X < Y, search(X, L), !. % search left subtree if X is less than Y
search(X, t(Y, _, R)) :- X >= Y, search(X, R), !. % search right subtree if X is greater than or equal to Y

% preorder traversal of a binary tree
preorder(nil). % if the tree is empty, do nothing
preorder(t(X, nil, nil)) :- write(X), write(" "), !. % if the tree has no children, write the root and return
preorder(t(X, L, R)) :- write(X), write(" "), preorder(L), preorder(R), !. % write the root, then traverse the left subtree, then traverse the right subtree

% postorder traversal of a binary tree
postorder(nil). % if the tree is empty, do nothing
postorder(t(X, nil, nil)) :- write(X), write(" "), !. % if the tree has no children, write the root and return
postorder(t(X, L, R)) :- postorder(L), postorder(R), write(X), write(" "),!. % traverse the left subtree, then traverse the right subtree, then write the root

% inorder traversal of a binary tree
inorder(nil). % if the tree is empty, do nothing
inorder(t(X, nil, nil)) :- write(X), write(" "), !. % if the tree has no children, write the root and return
inorder(t(X, L, nil)) :- inorder(L), write(X), write(" "), !. % if the tree has no right child, traverse the left subtree, then write the root
inorder(t(X, nil, R)) :- write(X), write(" "), inorder(R), !. % if the tree has no left child, write the root, then traverse the right subtree
inorder(t(X, L, R)) :- inorder(L), write(X), write(" "), inorder(R), !. % if the tree has both children, traverse the left subtree, then write the root, then traverse the right subtree
